### PR TITLE
Decorator for easier tool building

### PR DESCRIPTION
**Pull request by aymeric-roucher**
# What does this PR do?

We need an easier way to create tools than the heavy Tool class : but at the same time this class is useful since it's flexible. Se I just add a `@tool` decorator to make it easy to create a tool from a function, provided that it has the proper type hints and description.

### CODE REVIEW

**Comment by aymeric-roucher** [comment]
cc @Rocketknight1 
@LysandreJik I replace all input types "text" with "string" to conform with the common JSON schema type that we also use in `apply_chat_template`.

File: tests/agents/test_final_answer.py [hunk 1]
```diff
@@ -19,22 +19,20 @@
...
 
-from .test_tools_common import ToolTesterMixin
-
 
 if is_torch_available():
     import torch
 
 
-class FinalAnswerToolTester(unittest.TestCase, ToolTesterMixin):
+class FinalAnswerToolTester(unittest.TestCase):
```
Comment (LysandreJik) [inline tests/agents/test_final_answer.py:32]: Expected to remove the tool tester mixin here?

**Review by LysandreJik — APPROVED** [review]
:ok_hand:

File: tests/agents/test_final_answer.py (hunk 1 above)
Comment (aymeric-roucher) [reply tests/agents/test_final_answer.py]: I had removed it because FinalAnswerTool does not need the input/output type testing that other tools do need, but indeed we can keep it (and it may be useful if some tests are added to `ToolTes) so I added it back!

File: src/transformers/agents/tools.py [hunk 2]
```diff
@@ -808,3 +844,37 @@ def __init__(self, collection_slug: str, token: Optional[str] = None):
...
+
+    original_signature = inspect.signature(tool_function)
+    new_parameters = [inspect.Parameter("self", inspect.Parameter.POSITIONAL_OR_KEYWORD)] + list(
+        original_signature.parameters.values()
+    )
+    new_signature = original_signature.replace(parameters=new_parameters)
+    SpecificTool.forward.__signature__ = new_signature
+
+    SpecificTool.__name__ = class_name
+    return SpecificTool()
```
Comment (Rocketknight1) [inline src/transformers/agents/tools.py:880]: We discussed this a little on Slack but I've been thinking about it: Do you think users will get confused because of the different approaches taken in chat templates / agents?

Chat templates: Tools are passed as **JSON schema**. Users can also directly pass Python functions, and these will be automatically converted to JSON schema.

Agents: Tools are passed as `Tool` objects. Users can convert functions to `Tool` objects using the `@tool` decorator.

Do we need both approaches? We could have a more unified API by, for example, removing the `tool` decorator and instead converting functions to `Tool` objects if they're passed to agents, the same way we do in chat templates. That way, even if the internal objects are different, the API for users would be the same.

Comment (aymeric-roucher) [reply src/transformers/agents/tools.py:880]: Thank you for taking a look @Rocketknight1 !

So it seems to me that a fundamental difference is that in Chat templates, you can pass a JSON scheme : this is completely impossible for our tools since the JSON cannot encode the internal logic of the tool's forward pass.

But we can add the functionality to accept raw functions as tools, and only convert them to tools upon agent initialization! So we would support both `Tool` objects and passing raw functions.

I think we can keep supporting the decorator as well, since applying a decorator directly to the tool has the advantage to let the users try out the tool before it's passed to the agent initialization (which I personnally often do to make sure they work properly). Anyway this decorator is just the `tool` function, which we will need in any case for converting functions to Tools.

So I'd say let's keep the usage of `tool` as a decorator and add support for raw functions on agent initialization (which will also use `tool` behind the hood for conversion to `Tool` objects)! Wdyt as well @LysandreJik ?

Comment (aymeric-roucher) [reply src/transformers/agents/tools.py]: Ok let's merge this for now, since the solution you propose @Rocketknight1 only needs to add support for raw functions in the `tools` list upon agent initialization, which we can do in a later PR!

Comment (Rocketknight1) [reply src/transformers/agents/tools.py]: Sure!
